include/MidiApple.h

```diff
@@ -0,0 +1,153 @@
+/*
+ * MidiApple.h - Apple raw MIDI client
+ *
+ * Copyright (c) 2005-2015
+ *
+ * This file is part of LMMS - http://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef MIDI_APPLE_H
+#define MIDI_APPLE_H
+
+#include "lmmsconfig.h"
+
+#ifdef LMMS_BUILD_APPLE
+
+#include "MidiClient.h"
+#include "MidiPort.h"
+#include <CoreMIDI/CoreMIDI.h>
+
+
+class QLineEdit;
+
+
+class MidiApple : public QObject, public MidiClient
+{
+    Q_OBJECT
+public:
+	MidiApple();
+	virtual ~MidiApple();
+
+    static QString probeDevice();
+    
+    
+    inline static QString name()
+    {
+        return QT_TRANSLATE_NOOP( "setupWidget", "Apple MIDI" );
+    }
+    
+    
+    
+    virtual void processOutEvent( const MidiEvent & _me,
+                                 const MidiTime & _time,
+                                 const MidiPort * _port );
+    
+    virtual void applyPortMode( MidiPort * _port );
+    virtual void removePort( MidiPort * _port );
+    
+    
+    // list devices as ports
+    virtual QStringList readablePorts() const
+    {
+        return m_inputDevices.keys();
+    }
+    
+    virtual QStringList writablePorts() const
+    {
+        return m_outputDevices.keys();
+    }
+    
+    // return name of port which specified MIDI event came from
+    virtual QString sourcePortName( const MidiEvent & ) const;
+    
+    // (un)subscribe given MidiPort to/from destination-port
+    virtual void subscribeReadablePort( MidiPort * _port,
+                                       const QString & _dest,
+                                       bool _subscribe = true );
+    virtual void subscribeWritablePort( MidiPort * _port,
+                                       const QString & _dest,
+                                       bool _subscribe = true );
+    virtual void connectRPChanged( QObject * _receiver,
+                                  const char * _member )
+    {
+        printf("connectRPChanged called\n");
```

Comment:
I wouldn't worry too much about it, but yes - qDebug is preferable because
it's easier to hook.
On Jul 31, 2015 3:42 PM, "Dave" <notifications@github.com> wrote:

> In include/MidiApple.h
> <https://github.com/LMMS/lmms/pull/2228#discussion_r36023148>:
>
> > +    }
> > +
> > +    // return name of port which specified MIDI event came from
> > +    virtual QString sourcePortName( const MidiEvent & ) const;
> > +
> > +    // (un)subscribe given MidiPort to/from destination-port
> > +    virtual void subscribeReadablePort( MidiPort * _port,
> > +                                       const QString & _dest,
> > +                                       bool _subscribe = true );
> > +    virtual void subscribeWritablePort( MidiPort * _port,
> > +                                       const QString & _dest,
> > +                                       bool _subscribe = true );
> > +    virtual void connectRPChanged( QObject * _receiver,
> > +                                  const char * _member )
> > +    {
> > +        printf("connectRPChanged called\n");
>
> Yes, I believe we are. IIRC @Wallacoloo <https://github.com/Wallacoloo>
> posted on this subject.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LMMS/lmms/pull/2228/files#r36023148>.
>